src/oauth.ts

```diff
@@ -0,0 +1,505 @@
+    // Store the mapping between state and client's actual redirect URI with timestamp
+    if (params.state && params.redirectUri) {
+      try {
+        this.db.prepare('INSERT OR REPLACE INTO state_mappings (state, redirect_uri, timestamp) VALUES (?, ?, ?)')
+          .run(params.state, params.redirectUri, Date.now());
+        logger.debug(`Stored state mapping: ${params.state} -> ${params.redirectUri} at ${new Date().toISOString()}`);
+      } catch (err) {
+        logger.error('Error storing state mapping:', err);
+        throw err;
+      }
+    }
+
+    // Construct the authorization URL directly to ensure proper formatting
+    const authUrl = new URL(this._endpoints.authorizationUrl);
+
+    // Add required OAuth parameters
+    authUrl.searchParams.set('client_id', config.GITLAB_OAUTH2_CLIENT_ID!);
+    authUrl.searchParams.set('response_type', 'code');
+    authUrl.searchParams.set('redirect_uri', config.GITLAB_OAUTH2_REDIRECT_URL!.trim());
+    authUrl.searchParams.set('code_challenge', params.codeChallenge);
+    authUrl.searchParams.set('code_challenge_method', 'S256');
+
+    // Add optional parameters
+    if (params.state) {
+      authUrl.searchParams.set('state', params.state);
+    }
+
+
+    const gitlabScopes = ['api', 'openid','profile','email'];
+    authUrl.searchParams.set('scope', gitlabScopes.join(' '));
+
+    // GitLab doesn't support the 'resource' parameter, so we skip it
+
+    logger.debug({
+      url: authUrl.toString(),
+      scopes: gitlabScopes,
+      requested_scopes: params.scopes
+    },`Redirecting to GitLab OAuth`);
+
+    // Redirect to GitLab
+    res.redirect(authUrl.toString());
+  }
+
+  // Method to get redirect URI from state
+  async getRedirectUriFromState(state: string): Promise<{ redirectUri: string; timestamp: number } | undefined> {
+    const row = this.db.prepare('SELECT redirect_uri, timestamp FROM state_mappings WHERE state = ?').get(state) as {
+      redirect_uri: string;
+      timestamp: number;
+    } | undefined;
+
+    if (!row) {
+      return undefined;
+    }
+
+    return {
+      redirectUri: row.redirect_uri,
+      timestamp: row.timestamp
+    };
+  }
+
+  // Method to delete state mapping
+  async deleteStateMapping(state: string): Promise<void> {
+    this.db.prepare('DELETE FROM state_mappings WHERE state = ?').run(state);
+  }
+
+  // Method to verify token
+  async verifyToken(token: string): Promise<{ authInfo: AuthInfo; gitlabToken: string; timestamp: number } | undefined> {
+    // Get all token hashes to check against
+    const rows = this.db.prepare('SELECT * FROM access_tokens').all() as Array<{
+      token_hash: string;
+      client_id: string;
+      scopes: string;
+      expires_at: number | null;
+      timestamp: number;
+    }>;
+
+    // Find the matching token by verifying against each hash
+    let matchingRow = null;
+    for (const row of rows) {
```

Comment:
Hmm this does not seem optimal, do we not have another way to identify the correct token? Like an id?

Reply:
we can't store a deterministic id because that would be saving an unsalted hash in the database, which I don't want to do.

if there was a non-deterministic id, then we would need to somehow append it to the api key, and then extract it on read, but if i did that, then the token would no longer be a valid gitlab token

Reply:
in truth, we don't really need to do this at all. we can also

1. just pass it through, because the token will fail authentication to the gitlab api anyways
2. just check the api token with a request against the gitlab api every time, since it's supposed to just be a valid gitlab token anyways